server/cluster/cluster.go

```diff
@@ -40,40 +39,6 @@ func NewServer(db db.ArgoDB, enf *rbac.Enforcer, cache *servercache.Cache, kubec
 	}
 }
 
-func (s *Server) getConnectionState(cluster appv1.Cluster, errorMessage string) (appv1.ConnectionState, string) {
-	if clusterInfo, err := s.cache.GetClusterInfo(cluster.Server); err == nil {
-		return clusterInfo.ConnectionState, clusterInfo.Version
-	}
-	now := v1.Now()
-	clusterInfo := servercache.ClusterInfo{
-		ConnectionState: appv1.ConnectionState{
-			Status:     appv1.ConnectionStatusSuccessful,
-			ModifiedAt: &now,
-		},
-	}
-
-	config := cluster.RESTConfig()
-	config.Timeout = time.Second
-	version, err := s.kubectl.GetServerVersion(config)
-	if err != nil {
-		clusterInfo.Status = appv1.ConnectionStatusFailed
-		clusterInfo.Message = fmt.Sprintf("Unable to connect to cluster: %v", err)
-	} else {
-		clusterInfo.Version = version
-	}
-
-	if errorMessage != "" {
-		clusterInfo.Status = appv1.ConnectionStatusFailed
-		clusterInfo.Message = fmt.Sprintf("%s %s", errorMessage, clusterInfo.Message)
-	}
-
-	err = s.cache.SetClusterInfo(cluster.Server, &clusterInfo)
-	if err != nil {
-		log.Warnf("getClusterInfo cache set error %s: %v", cluster.Server, err)
-	}
-	return clusterInfo.ConnectionState, clusterInfo.Version
-}
-
 // List returns list of clusters
 func (s *Server) List(ctx context.Context, q *cluster.ClusterQuery) (*appv1.ClusterList, error) {
 	clusterList, err := s.db.ListClusters(ctx)
```

Comment:
The `clustersByServer` map , and `kube.RunAllAsync` were required because state check is time consuming. So we had to run checks in parallel for each cluster. 

Now we can get rid of map and use normal for loop to reduct secrets from clusters and apply RBAC checks.